src/shift.rs

```diff
@@ -1,231 +1,584 @@
-                    })
-            }
-            _ => None,
-        },
-    )
-}
-
-fn match_node(node: &str, pattern: &str) -> bool {
-    if pattern == "*" {
-        return true;
+}
+
+fn apply<'ctx, 'input: 'ctx>(
+    obj: &'input Obj,
+    path: &'ctx mut Vec<(Vec<Cow<'input, str>>, &'input Value)>,
+    out: &'ctx mut Value,
+) -> Result<()> {
+    match_obj_and_key_impl(
+        obj,
+        path,
+        path.last().unwrap().0[0].clone(),
+        path.last().unwrap().1,
+        out,
+        LhsSelection::Infallible,
+    )?;
+    let input = path.last().unwrap();
+
+    match input.1 {
+        Value::Object(input) => {
+            for (k, v) in input.iter() {
+                match_obj_and_key(obj, path, Cow::Borrowed(k), v, out)?;
+            }
+        }
+        Value::Bool(b) => {
+            let k = if *b { "true" } else { "false" };
+
+            match_obj_and_key(obj, path, Cow::Borrowed(k), input.1, out)?;
+        }
+        Value::Array(arr) => {
+            for (k, v) in arr.iter().enumerate() {
+                let k = k.to_string();
+                match_obj_and_key(
+                    obj,
+                    path,
+                    // this makes the downstream functions to do some extra allocations.
+                    // could avoid some of these allocations by mapping some small indexes to static str's
+                    Cow::Owned(k),
+                    v,
+                    out,
+                )?;
+            }
+        }
+        Value::Number(n) => {
+            let k = n.to_string();
+
+            match_obj_and_key(obj, path, Cow::Owned(k), input.1, out)?;
+        }
+        Value::String(k) => {
+            match_obj_and_key(obj, path, Cow::Borrowed(k), input.1, out)?;
+        }
+        Value::Null => {
+            let k = "null";
+            match_obj_and_key(obj, path, Cow::Borrowed(k), input.1, out)?;
+        }
+    };
+
+    Ok(())
+}
+
+fn match_obj_and_key<'ctx, 'input: 'ctx>(
+    obj: &'input Obj,
+    path: &'ctx mut Vec<(Vec<Cow<'input, str>>, &'input Value)>,
+    k: Cow<'input, str>,
+    v: &'input Value,
+    out: &'ctx mut Value,
+) -> Result<()> {
+    if match_obj_and_key_impl(obj, path, k.clone(), v, out, LhsSelection::Literal)? {
+        return Ok(());
     }
-    if node == pattern {
-        return true;
+    if match_obj_and_key_impl(obj, path, k.clone(), v, out, LhsSelection::Amp)? {
+        return Ok(());
     }
-    if pattern.split('|').any(|part| part == node) {
-        return true;
+    if match_obj_and_key_impl(obj, path, k.clone(), v, out, LhsSelection::Pipes)? {
+        return Ok(());
     }
-    false
+
+    Ok(())
+}
+
+fn lhs_is_fallible(lhs: &Lhs) -> bool {
+    !matches!(lhs, Lhs::DollarSign(_, _) | Lhs::Square(_) | Lhs::At(_))
+}
+
+#[derive(PartialEq)]
+enum LhsSelection {
+    Infallible,
+    Literal,
+    Amp,
+    Pipes,
 }
 
-#[cfg(test)]
-mod test {
+fn match_obj_and_key_impl<'ctx, 'input: 'ctx>(
```

Comment:
seems like this needs unit test

Reply:
imo, testing it becomes kind of a integration test because it calls other functions which call other functions and we have integration tests already.

Reply:
It isn't easy to understand logic without a unit test

Reply:
Is it ok If I add comment descriptions instead of unit tests for these?

Reply:
let's add comments but it would be good have internal unit tests

Reply:
Thanks. adding comments now, and I'll try to add unit tests for self contained functions in future PR